pkgs/top-level/all-packages.nix

```diff
@@ -12206,39 +12206,47 @@ with pkgs;
   fastStdenv = overrideCC gccStdenv (wrapNonDeterministicGcc gccStdenv buildPackages.gcc10);
 
   wrapCCMulti = cc:
-    if stdenv.targetPlatform.system == "x86_64-linux" then let
-      # Binutils with glibc multi
-      bintools = cc.bintools.override {
-        libc = glibc_multi;
-      };
-    in lowPrio (wrapCCWith {
-      cc = cc.cc.override {
-        stdenv = overrideCC stdenv (wrapCCWith {
-          cc = cc.cc;
-          inherit bintools;
-          libc = glibc_multi;
+    if stdenv.targetPlatform.system != "x86_64-linux" then
+      throw "Multilib ${cc.name} not supported for ‘${stdenv.targetPlatform.system}’"
+    else
+      let
+        gccForLibs = callPackage ../development/compilers/llvm/multi-gcc-libs.nix {
+          gcc32 = pkgsi686Linux.gcc;
+          gcc64 = pkgs.gcc;
+        };
+        libc =
+          /**/ if cc.isGNU   or false then glibc_multi_gcc
+          else if cc.isClang or false then glibc_multi_clang
+          else throw "unknown C compiler, cannot make multilib wrapper";
+        bintoolsMulti = cc.bintools.override {
+          inherit libc;
+        };
+      in
+        lowPrio (wrapCCWith {
```

Comment:
I believe we are overwriting some of `clang`'s attributes when creating a CC from scratch through `wrapCCWith`.

I'm still thinking of a cleaner way to achieve this, but this small PoC solves a clang's inability to find its own stdlib includes:

```
diff --git a/pkgs/top-level/all-packages.nix b/pkgs/top-level/all-packages.nix
index 65fac823e68..92ceb70ddbc 100644
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -12222,6 +12222,7 @@ with pkgs;
           inherit libc;
         };
       in
+        if cc.isGNU then
         lowPrio (wrapCCWith {
           inherit libc gccForLibs;
 
@@ -12242,7 +12243,11 @@ with pkgs;
           extraBuildCommands = ''
             echo "dontMoveLib64=1" >> $out/nix-support/setup-hook
           '';
-        });
+        }) else
+        cc.override {
+          inherit libc gccForLibs;
+          bintools = bintoolsMulti;
+        };
 
 
   gcc_multi = wrapCCMulti gcc;
```

This fixes the includes but `clang`'s still cannot find `libgcc`